solana/solana-ibc/programs/solana-ibc/src/validation_context.rs

```diff
@@ -79,27 +81,40 @@ impl ibc::ValidationContext for IbcStorage<'_, '_> {
         height: &ibc::Height,
     ) -> Result<Self::AnyConsensusState> {
         let store = self.borrow();
-        let state = if height.revision_number() == 1 {
-            store.chain.consensus_state(height.revision_height().into())?
-        } else {
-            None
+        #[cfg(feature = "witness")]
+        {
+            let (_slot, fetched_timestamp, fetched_trie_root) = store
```

Comment:
Should we check height.revision_number here as well?

Reply:
yes i think we should.

Reply:
It might be easier to define a helper method which returns
`Result<Option<Self::AnyConsensusState>>` and have this use that:

        fn host_consensus_state(
            &self,
            height: &ibc::Height,
        ) -> Result<Self::AnyConsensusState> {
            self.host_consensus_state_impl(height).and_then(|state| {
                state.ok_or(ibc::ClientError::MissingLocalConsensusState {
                    height: *height,
                })
            })
        }

    // ...

        #[cfg(feature = "witness")]
        fn host_consensus_state_impl(
            &self,
            height: &ibc::Height,
        ) -> Result<Option<Self::AnyConsensusState>> {
            if height.revision_number() != 0 {
                return Ok(None);
            }
            Ok(self.borrow()
                .private
                .local_consensus_state
                .iter()
                .find(|cs| cs.0 == height.revision_height())
                .map(|(_slot, timestamp, trie_root)| {
                    AnyConsensusState::Rollup(cf_solana::ConsensusState {
                        trie_root: trie_root.to_vec().into(),
                        timestamp_sec: NonZeroU64::new(*timestamp).unwrap(),
                    })
                }))
        }

        #[cfg(not(feature = "witness"))]
        fn host_consensus_state_impl(
            &self,
            height: &ibc::Height,
        ) -> Result<Option<Self::AnyConsensusState>> {
            if height.revision_number() != 1 {
                return Ok(None);
            }
            let height = height.revision_height().into();
            self.borrow().chain.consensus_state(height).map(|state| {
                let state = cf_guest::ConsensusState {
                    block_hash: state.0.to_vec().into(),
                    timestamp_ns: state.1,
                };
                Some(AnyConsensusState::Guest(state))
            })
        }

Again, up to you.

Reply:
Hmm, actually, since we check for revision_number == 1 in both cases, that could be moved outside of the _impl method as well. That would potentially eliminate use of Option::then.